### PR TITLE
[Seq] Add a clock multiplexer op

### PR DESCRIPTION
**Pull request by nandor**
As part of the introduction of a clock type (#5851), this PR adds a clock multiplexer which later will operate on `seq.clock`.

### CODE REVIEW

**Comment by teqdruid** [comment]
The title for this is wrong.

**Comment by nandor** [comment]
Unfortunately the edit distance to `multiplier` was less than the edit distance to `multiplexer`.

**Comment by mortbopet** [comment]
I might be missing something obvious, but what is "safe" about this, seeing as the provided lowering is just a `comb.mux` anyways?

**Comment by nandor** [comment]
The intention is to introduce a clock type for this op to operate upon, in the meantime it is pretty much sugar on top of `comb.mux`.

**Comment by teqdruid** [comment]
I'd imagine it'll be lowered to a special ASIC std cell / construct which does clock multiplexing in a safe manner.

**Comment by sequencer** [comment]
That's great! Also looking forward to `clkxor` and other clock cells for intmodule.

**Comment by nandor** [comment]
Lowering to `BUFGCTRL` might be an option. But for now, it might be valuable to have a special representation that, on top of `comb.mux`, allows us to to target the op in auto-generated metadata, such as XDC constraints.

File: include/circt/Dialect/Seq/SeqOps.td [hunk 1]
```diff
@@ -324,6 +324,32 @@ def ClockGateOp : SeqOp<"clock_gate", [
...
+//===----------------------------------------------------------------------===//
+// Clock Multiplexer
+//===----------------------------------------------------------------------===//
+
+def ClockMuxOp : SeqOp<"clock_mux", [Pure]> {
+  let summary = "Safely selects a clock based on a condition";
+  let description = [{
+    The `seq.clock_mux` op selects a clock from two options based on a
+    condition without glitches. If `cond` is true, the first clock operand
+    is selected to drive downstream logic. Otherwise, the second clock is used.
```
Comment (fabianschuiki) [inline include/circt/Dialect/Seq/SeqOps.td:336]: I would remove the `without glitches` part.

In ASIC technologies, the dedicated clock mux cells are almost always just regular mux cells, but with the 0-to-1 rise and 1-to-0 fall time of their output balanced, while regular mux cell outputs fall faster than they rise. (PMOS transistors are generally weaker/slower than NMOS; ASIC stdcells have larger PMOS than NMOS to compensate a bit; regular muxes don't compensate fully, so rise time is usually slower than fall time; clock muxes trade cell area for providing balanced rise and fall times.)

To provide glitchless muxing, you'd generally combine a clock mux with a clock gate cell and have a state machine that first gates the clock, switches the mux once it's safe, and then ungates it again. Same goes for clock XOR/AND/OR gates. Mostly about balanced rise/fall times.

**Review by fabianschuiki — APPROVED** [review]
I like the idea of having a well-known set of operations that are allowed on a clock. I've seen gating, muxing, AND, OR, and XOR in ASIC stdcell libraries in the wild.

Stuff like arcilator or some other passes could then try to be clever and fold those clock ops into the enable conditions of registers, or pull enable conditions out of registers and implementt them using these clock ops.

**Pull request merged**